---
title: "Check DADA2-formatted Silva 138.1 SSU taxonomy files"
author: "Michael McLaren"
date: "2021-03-07"
---

```{r}
library(tidyverse)

packageVersion("dada2")

silva_version <- "138.1"

path <- here::here()
```

```{r}
output <- c(
  train = "silva_nr99_v{silva_version}_train_set.fa.gz",
  train_with_species = "silva_nr99_v{silva_version}_wSpecies_train_set.fa.gz",
  species = "silva_species_assignment_v{silva_version}.fa.gz"
) %>%
  map_chr(str_glue) %>%
  map_chr(~file.path(path, "output", .))
```

Make sure we can read in each database and it looks as expected,
```{r}
x <- output %>%
  map(Biostrings::readDNAStringSet)
x
```

Breakdown of included NR99 sequences by kingdom:
```{r}
x[[1]] %>% names %>% str_extract("^[^;]+") %>% table
```

Note the small number of Eukaryote sequences. 
The current DADA2 formatting function only keeps a small set of Eukaryote sequence for use as an outgroup, and these files should not be used for eukaryotic 18S sequence analysis.

```{r}
rm(x)
```

## Test assignment on V4 gut microbiome data

I will use V4 amplicon data of animal fecal microbiomes from Thanissery et al (2019) doi:10.1101/822577, available at https://github.com/rthanis/animal-cdiff and archived at https://zenodo.org/record/3691924. 
I will download a phyloseq object that contains ASV sequences in the `refseq` slot and taxonomic assignments in the `tax_table` slot that are against the previous Silva release (v132), made using DADA2's taxonomy assignment functions.
```{r}
library(phyloseq)
library(tidyverse)

ps <- readRDS(
  url("https://github.com/rthanis/animal-cdiff/raw/anaerobe/results/ps.Rds")
)
ps
```
To speed things up I'll use a subset of ASVs,
```{r}
ps0 <- prune_taxa(taxa_names(ps)[c(1:100, 1000:1150, 10300:10343)], ps)
ps0
```
And assign ASV sequences to the new Silva database files, allowing for multiple species assignments as done for the original dataset.
```{r}
tax.new <- ps0 %>%
  refseq %>%
  dada2::assignTaxonomy(
    refFasta = output["train"],
    multithread = TRUE
  ) %>%
  dada2::addSpecies(
    refFasta = output["species"],
    allowMultiple = TRUE
  )
```
Combine with the existing Silva 132 assignments to compare results,
```{r}
tb0 <- tax_table(ps0) %>%
  as("matrix") %>%
  as_tibble %>%
  select(-NCBI_species)
tb1 <- as_tibble(tax.new)
all.equal(tb0$Kingdom, tb1$Kingdom)
tb <- bind_cols(
  # Original assignments against Silva 132
  tb0 %>% dplyr::rename_with(~str_c(., ".0")), 
  # Assignments against new Silva
  tb1 %>% dplyr::rename_with(~str_c(., ".1")),
)
check <- tb %>%
  select(starts_with("Phylum"), starts_with("Genus"))
```
```{r}
head(check, 10)
tail(check, 10)
```

Do the same for species assignments,
```{r}
check_species <- tb %>%
  select(starts_with("Species"))
head(check_species, 10)
tail(check_species, 10)
```

## Session info

```{r, R.options = list(width = 83)}
sessioninfo::session_info()
```
